lib/jobs/ucs-discovery.js

```diff
@@ -186,15 +151,114 @@ function UcsDiscoveryJobFactory(
+                                return self.createNode(data, Constants.NodeTypes.Compute)
+                                .spread(function(newNode) {
+                                    var relations = [{
+                                                     relationType: 'enclosedBy',
+                                                     targets: [chassisId]
+                                                 }];
+                                    nodeList.push(newNode.id);
+                                    return self.updateRelations(newNode.id, relations);
+                                });
+                            })
+                            .then(function() {
+                                var chassisRelations = [{
+                                    relationType: 'encloses',
+                                    targets: nodeList
+                                }];
+                                return self.updateRelations(chassisId, chassisRelations);
+                            });
+                        });
+                });
+        });
+    };
+
+    UcsDiscoveryJob.prototype.updateRelations = function(nodeId, relations) {
+        // Update existing node with new relations or create one
+        return waterline.nodes.needOneById(nodeId)
+            .then(function(curNode) {
+                relations = _.uniq(relations.concat(curNode.relations), 'relationType');
+                return waterline.nodes.updateOne(
+                    { id: curNode.id },
+                    { relations: relations }
+                );
+            });
+    };
+
+
+    UcsDiscoveryJob.prototype.createNode = function(data, type) {
+
+        var self = this;
+        var config = Object.assign({}, self.settings);
+
+        assert.string(data.name);
+        assert.string(data.path);
+
+        var id;
+        if (type === Constants.NodeTypes.Compute) {
+            assert.array(data.macs);
+            id = data.macs;
+        } else {
+            id = [config.ucs, data.path];
+        }
+
+        var obm = {
+            config: config,
+            service: 'ucs-obm-service'
+        };
+
+        var node = {
+            type: type,
+            name: data.path,
+            identifiers: id,
+            relations : []
+        };
+
+        // Update existing node  or create a new one
+        return waterline.nodes.needOne({identifiers:data.macs})
+            .then(function(curNode) {
+                return waterline.nodes.updateOne(
+                    { id: curNode.id },
+                    node
+                );
+            })
+            .catch(function(error) {
+                if (error.name === 'NotFoundError') {
+                    return waterline.nodes.create(node);
+                }
+                throw error;
+            })
+            .then(function(data){
+                return [data, waterline.obms.upsertByNode(data.id,obm)];
```

Comment:
@geoff-reid It doesn't look like you ever use the fulfillment value from the obms upsert, so maybe you can just do this instead:
```
.tap(function(data) {
    return waterline.obms.upsertByNode(data.id,obm);
});
```
That will return a promise that fulfills to `data` when the upsert promise fulfills..